### ISSUE TITLE
ChainSource::Prompt should support a pick-list/enum style variant

### ISSUE THREAD

**Issue by anussel5559** [LucasPickering/slumber#352]
**Did you [search](https://github.com/LucasPickering/slumber/issues) for existing issues already?** Yes

**Is your feature request related to a problem? Please describe.**
Not a problem, more a QoL feature

**Describe the solution you'd like**
I'm probably stepping in to a wonderfully sticky tar pit here, but I've run in to a few scenarios while using slumber that could be significantly enhanced by the ability to use a pick-select on a prompt ChainSource - effectively an enum variant of a prompt that is driven by a simple text array, or another chain (that returns an array)

Some specifics: I have an API ecosystem where a list of resources returned from a list-style route can inform the id used to perform subsequent singular-resource operations (update, get more details, delete). On these subsequent singular-resource operations, I'd love to get prompted to choose one of the resources previously returned from the list-style route.

Ideally, I'd like to NOT place this resource id in a profile as the profile might be committed and shared with others as part of the API service. And trying a copy-paste workflow to solve this scenario could get old quick (and breaks the flow while using slumber, which I really enjoy).

**Describe alternatives you've considered**
Nothing really can replicate this experience

**Additional context**
None

**Comment by LucasPickering** [comment]
I like this suggestion. Just to have something concrete to talk about, I'm picturing YAML like this:

```yaml
chains:
  select:
    source: !select
      message: Pick an option
      options:
        - option 1
        - option 2
```

The `message` field, and each option in `options`, would be a template and therefore dynamic. This seems pretty useful and achieveable, but I see one main issue here: You need a static number of options in the list. Each individual option is dynamic, but the list as a whole is static length. It sounds like you want to be able to generate the entire list dynamically from a template, which sounds very useful but requires a lot more consideration.

Currently, templates only render to strings (for bodies and headers they can also render to non-UTF-8 bytes). In order to have a template render to a list, we need to define some sort of list syntax. For consumers that want a list rather than a string, they would render to a string then perform extra parsing to get a list. I see two obvious choices for list syntax:
- CSV
  - This would be simple to write by hand, but that's probably not relevant since the goal here is to generate lists dynamically
- JSON lists
  - This would be easier to generate dynamically, e.g. if you want to build a list from a chained response, you could just use JSONPath to spit out a list
  - Has the added benefit that we already have a parser built for this

Here's a full example of what I'm thinking. In this case, assume the `/users` endpoint returns a response like:
```json
{
  "count": 3,
  "results": [
    {"id": 1, "name": "Mike"},
    {"id": 2, "name": "Anastasia"},
    {"id": 3, "name": "Markus"}
  ]
}
```

Here's what the collection would look like:

```yaml
chains:
  user_ids:
    source: !request
      recipe: list
    selector: $.results[*].id
  user_id:
    source: !select
      options: "{{chains.user_ids}}"

requests:
  list: !request
    method: GET
    url: "{{host}}/users"
  get: !request
    method: GET
    url: "{{host}}/users/{{chains.user_id}}"
```

> **Note:** `options` would now accept a template OR an array of options, so you could still provide a static list if you want

In this case, when you make the `get` request, it would open a popup to pick from the options of 1, 2, or 3 for user ID. Some additional edge cases to consider:
- If the template for `options` renders to something other than a JSON list (e.g. just a string, a JSON object, etc.), we should return an error because there's no way to get a list out of that
- If the template renders to a list, but the list contains some non-string values (e.g. `[1, "string", {"a": 1}]`), I would just stringify all the values. So in this case the three options would be `"1"`, `"string"`, and `"{\"a\": 1}"`

What are your thoughts on this design? Does this solve your problem? I don't love the complexity of the list parsing, but since I don't see an alternative to accomplish this right now, I think it's worth it.

If you're interested in working on implementation, I'll warn you that I'm planning on releasing 2.0 in the next couple weeks so I don't want to merge any big features until then. So you can start on this but it would be a bit until it's released. The sequence of steps I would follow to implement this are:
- Add the `!select` variant but only support static lists
  - Add a `Select` variant to [ChainSource](https://github.com/LucasPickering/slumber/blob/4ab1ee42812148b1427090b343d21292ca235713/crates/slumber_core/src/collection/models.rs#L396)
    - For now the `options` field would just be a `Vec<Template>`
  - Add a method to render the template [here](https://github.com/LucasPickering/slumber/blob/4ab1ee42812148b1427090b343d21292ca235713/crates/slumber_core/src/template/render.rs#L342)
    - Since this has to pause rendering to ask the user for a response, you can add another method to the [Prompter trait](https://github.com/LucasPickering/slumber/blob/master/crates/slumber_core/src/template/prompt.rs), that sends a list of values and receives an index back (corresponding to a particular value).
      - To start you can just add a stub implementation with the `unimplemented!` macro to the TUI and CLI implementations
- Link the prompter up to the CLI and TUI frontends
  - TUI should pop up a modal
  - CLI should show a select list using `dialoguer`
- Add support for dynamic lists
  - Make the `options` field an enum that takes either `Vec<Template>` or `Template`
  - Update the rendering implementation in `render.rs` to handle the `Template` case. It should render the template, then parse it with `serde_json::from_str`, and apply the logic I described above to ensure we have a list of strings
- Documentation

Some additional things to consider:
- Do NOT try to do this all in one PR. It'll harder to write and harder to review. The 4 top-level bullet points would probably break out into 4 PRs nicely, but feel free to organize it differently if you want
- Tests are important, especially in the `slumber_core` crate. TUI and CLI behavior isn't as important to test
- You don't have to commit to doing this whole thing. If you do the first chunk of work and decide you hate it, I can finish it

If you have any questions, feel free to email me (it's in my github profile) or I just created a [channel on gitter.im](https://matrix.to/#/#slumber-rs:gitter.im).

**Comment by anussel5559** [comment]
Awesome, yea, that's exactly how I was thinking about this!

I really appreciate the guidance/direction pointing here as well. I'm not worried about timing of release, just happy to help implement! I'll reach out if I hit any roadblocks along the way!

**Comment by anussel5559** [comment]
This is fully implemented (targeting v2.1)!
